Add release notes for 3.1 (#15238)
grafana/mimir: docs/sources/mimir/release-notes/v3.1.md

### More Kafka options for Ingest storage

Ingest storage now supports additional ways to authenticate with Kafka through the new `-ingest-storage.kafka.sasl-mechanism` flag, including SCRAM, OAUTHBEARER, and AWS MSK IAM authentication. In addition, new `-ingest-storage.kafka.tls*` flags allow connecting to Kafka clusters over TLS, including mTLS.

You can also configure multiple Kafka seed brokers via comma-separated values in `-ingest-storage.kafka.address` and enable rack-aware consumption with `-ingest-storage.kafka.client-rack`.

### Separate ingestion limits by tenant metadata

Distributors can now track limits separately based on tenant metadata. This allows operators to track limits separately for subsets of write requests belonging to the same tenant, for example to prioritize some sources of metrics over others.

Clients may pass tenant metadata in the `X-Scope-OrgID` header using the format `tenantID:key1=value1:key2=value2`, and operators may define per-metadata overrides in the runtime configuration.

### Mimir Query Engine (MQE) improvements

MQE continues to receive significant optimizations in this release:

> - Experimental extended range selector modifiers `smoothed` and `anchored`, enabled with `-query-frontend.enabled-promql-extended-range-selectors`.

Reviewer: ```suggestion
- Support for experimental PromQL extended range selector modifiers `smoothed` and `anchored`, enabled with `-query-frontend.enabled-promql-extended-range-selectors=true`.
```

- Optimization passes for common subexpression elimination, subset selector elimination, projection pushdown, and multi-aggregation without buffering.

> - Improved per-query memory consumption limit enforcement in histogram and aggregation operations.

Reviewer: ```suggestion
- Improved per-query memory consumption limit enforcement in a variety of scenarios.
```

- Experimental support for splitting and caching intermediate results for functions over range vectors in instant queries.

> - Experimental support for the `info()` PromQL function.

Reviewer: ```suggestion
- Support for the experimental `info()` PromQL function.
```

## Important changes

Grafana Mimir 3.1 introduces several updates that change default behavior and configuration. Review these changes before upgrading:

- Experimental support for disabling ring heartbeats and heartbeat timeouts has been removed.
- The `-target=flusher` mode has been removed; use the `/ingester/flush` HTTP endpoint instead.
- Uploaded TSDB blocks must now use v2 of the index file format. Store-gateways no longer generate index-headers from v1 index format blocks.
- Per-step stats are no longer supported when MQE is enabled. The `-query-frontend.cache-samples-processed-stats` flag is deprecated and has no effect.
- The `-querier.response-streaming-enabled` flag has been removed; active series responses are now always streamed.
- `cortex_ingest_storage_writer_buffered_produce_bytes` has been renamed to `cortex_ingest_storage_writer_buffered_produce_bytes_distribution`.
- Metric `cortex_ingester_owned_target_info_series` has been removed.
- The `cost_attribution_labels` configuration option has been removed; use `cost_attribution_labels_structured` instead.

> - The `minimum_step_size` filter in `blocked_queries` has been renamed to `step_size_shorter_than`. Blocked queries configuration is now validated at load time.

Reviewer 2: `minimum_step_size` was never available in a tagged release so we could probably skip this?

> Blocked queries configuration is now validated at load time.

This is the main change and affects both config reloading and process startup.
See (next docs):  https://grafana.com/docs/mimir/next/configure/configure-blocked-queries/#troubleshoot-invalid-configuration

Author: Thanks, see [94ed51e](https://github.com/grafana/mimir/pull/15238/commits/94ed51eb363daabfdc62e010b0fb2ef953471478)

- `-querier.prefer-availability-zone` has been renamed to `-querier.prefer-availability-zones` and now accepts a comma-separated list.

> - The per-query memory consumption limit now considers buffered messages and spans all time-split sub-queries when MQE is enabled.

Reviewer: It might be clearer to say this, as the details likely won't make sense to those not familiar with Mimir's internals:

```suggestion
- The per-query memory consumption limit now considers more sources of memory consumption. As a result, queries that previously succeeded may now fail due to exceeding the memory consumption limit.
```

- The following flags have been removed:
  - `-distributor.metric-relabeling-enabled`
  - `-compactor.no-blocks-file-cleanup-enabled`
  - `-compactor.in-memory-tenant-meta-cache-size`
  - `-blocks-storage.bucket-store.index-header.eager-loading-startup-enabled`
  - `*.memcached.dns-ignore-startup-failures`

## Experimental features

Grafana Mimir 3.1 includes some features that are experimental. Use these features with caution and report any issues that you encounter:

- New usage-tracker component to enforce series limits before data is ingested.
- Zone-aware memberlist routing to reduce cross-AZ data transfer.
- Query planning in query-frontends with distributed execution across queriers.
- MQE extended range selector modifiers (`smoothed`, `anchored`).
- MQE `info()` function.

> - MQE optimization passes: projection pushdown, multi-aggregation, subset selector elimination, common subexpression elimination for range vector expressions.

Reviewer: ```suggestion
- Support in MQE for experimental PromQL extended range selector modifiers (`smoothed`, `anchored`).
- Support in MQE for the experimental `info()` PromQL function.
- MQE optimization passes: multi-aggregation, subset selector elimination, common subexpression elimination for range vector expressions.
```
